src/stream/src/executor/aggregation/agg_impl/foldable.rs

```diff
@@ -1,1024 +0,0 @@
-        )
-        .unwrap();
-
-        assert_eq!(agg.get_output().unwrap().unwrap().as_float64(), &1.0);
-
-        agg.apply_batch(
-            &[Op::Insert, Op::Insert, Op::Insert, Op::Insert],
-            None,
-            &[&F64Array::from_iter([Some(1.0), Some(10.0), Some(-1.0), Some(5.0)]).into()],
-        )
-        .unwrap();
-        assert_eq!(agg.get_output().unwrap().unwrap().as_float64(), &-1.0);
-    }
-
-    #[test]
-    fn test_maximum() {
-        let mut agg = TestStreamingMaxAgg::<I64Array>::default();
-        agg.apply_batch(
-            &[Op::Insert, Op::Insert, Op::Insert, Op::Insert],
-            None,
-            &[&I64Array::from_iter([Some(10), Some(1), None, Some(5)]).into()],
-        )
-        .unwrap();
-
-        assert_eq!(agg.get_output().unwrap().unwrap().as_int64(), &10);
-
-        agg.apply_batch(
-            &[Op::Insert, Op::Insert, Op::Insert, Op::Insert],
-            None,
-            &[&I64Array::from_iter([1, 10, 100, 5]).into()],
-        )
-        .unwrap();
-        assert_eq!(agg.get_output().unwrap().unwrap().as_int64(), &100);
-    }
-
-    #[test]
-    fn test_bit_xor() {
-        let mut agg = TestStreamingBitXorAgg::<I64Array>::default();
-        agg.apply_batch(
-            &[Op::Insert, Op::Insert, Op::Insert, Op::Insert],
-            None,
-            &[&I64Array::from_iter([Some(10), Some(1), None, Some(5)]).into()],
-        )
-        .unwrap();
-
-        assert_eq!(agg.get_output().unwrap().unwrap().as_int64(), &14);
-
-        agg.apply_batch(
-            &[Op::Delete, Op::Delete, Op::Delete, Op::Delete],
-            None,
-            &[&I64Array::from_iter([1, 10, 100, 5]).into()],
-        )
-        .unwrap();
-        assert_eq!(agg.get_output().unwrap().unwrap().as_int64(), &100);
-    }
-
-    fn bench_i64(
-        b: &mut Bencher,
-        mut agg: Box<dyn StreamingAggImpl>,
-        agg_desc: &str,
-        chunk_size: usize,
-        vis_rate: f64,
-        iter_count: usize,
-        append_only: bool,
-    ) {
-        println!(
-            "benching {} agg, chunk_size {}, vis_rate {}, iter_count {}",
-            agg_desc, chunk_size, vis_rate, iter_count
-        );
-        let bitmap = if vis_rate < 1.0 {
-            Some(rand_bitmap::gen_rand_bitmap(
-                chunk_size,
-                (chunk_size as f64 * vis_rate) as usize,
-                666,
-            ))
-        } else {
-            None
-        };
-        let (ops, data) = rand_stream_chunk::gen_legal_stream_chunk(
-            bitmap.as_ref(),
-            chunk_size,
-            append_only,
-            666,
-        );
-        b.iter(|| {
-            for _ in 0..iter_count {
-                agg.apply_batch(&ops, bitmap.as_ref(), &[&data]).unwrap();
-            }
-        });
-    }
-
-    // TODO: refactor with macro
-    #[bench]
-    fn bench_foldable_sum_agg_without_vis(b: &mut Bencher) {
```

Comment:
would you help to migrate and run these benchmarks?

Reply:
They have been migrated to `src/expr/src/agg/general.rs`. I'm running these benches and will show the results soon.

Reply:
The result is shocking. 🤯

  | Before (ns) | After(ns) | Boost(Nx)
-- | -- | -- | --
count_agg_vis_rate_0_05 | 6,021 | 796 | 8
count_agg_vis_rate_0_25 | 32,725 | 962 | 34
count_agg_vis_rate_0_5 | 72,935 | 1,249 | 58
count_agg_vis_rate_0_75 | 110,661 | 1,491 | 74
count_agg_without_vis | 78,730 | 1,561 | 50
max_agg_vis_rate_0_05 | 12,234 | 791 | 15
max_agg_vis_rate_0_25 | 60,977 | 1,018 | 60
max_agg_vis_rate_0_5 | 122,474 | 1,328 | 92
max_agg_vis_rate_0_75 | 182,863 | 1,615 | 113
max_agg_without_vis | 146,554 | 1,724 | 85
min_agg_vis_rate_0_05 | 12,241 | 795 | 15
min_agg_vis_rate_0_25 | 60,960 | 1,019 | 60
min_agg_vis_rate_0_5 | 121,962 | 1,327 | 92
min_agg_vis_rate_0_75 | 182,723 | 1,615 | 113
min_agg_without_vis | 146,539 | 1,723 | 85
sum_agg_vis_rate_0_05 | 15,333 | 852 | 18
sum_agg_vis_rate_0_25 | 66,860 | 1,439 | 46
sum_agg_vis_rate_0_5 | 131,628 | 2,365 | 56
sum_agg_vis_rate_0_75 | 200,731 | 3,065 | 65
sum_agg_without_vis | 283,170 | 3,418 | 83

Reply:
Oops, it's not right. The old benches run 100 times in each iteration. 😅

Reply:
So exciting. :exploding_head:

Reply:
Fixed and updated result. The tide is turned. I should fix the regression now. 🥵

Reply:
I think it is because of the slow iteration on bitmap

Reply:
https://github.com/risingwavelabs/risingwave/pull/9310

Reply:
Right. It should use `iter_ones` in `update_multi`.

Reply:
Result updated after using `iter_ones`.

A notable overhead is the `start_row_id..end_row_id` range filter in `update_multi`. This is only used for batch agg (hashagg: update single row at a time, sortagg: update a range at a time). Other usages are all full ranges. I'm considering whether we should remove this range from the `update` method, and refactor batch usages by setting visibilities.

Reply:
> I'm considering whether we should remove this range from the update method, and refactor batch usages by setting visibilities.

+1 if performance is ok. This also helps increase code consistency.

Reply:
But the performance is a concern. Rewriting `update_single` to `update_multi` with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

Reply:
Anyway, do you think the performance shown above is acceptable to get this PR merged? We can leave further optimization to the future.

Reply:
> Rewriting `update_single` to `update_multi` with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

I think we can get rid of `update_single` finally, do batch updating in batch hash agg just like in streaming, but not necessarily in this PR.

Reply:
> But the performance is a concern. Rewriting update_single to update_multi with one bit set in visibility will definitely hurt performance. Perhaps we should first make batch hashagg update a group at a time (like streaming does).

+1 for this.
maybe we should do different interfaces for the predicate. one is for range, one is for visibity